### PR TITLE
Add the graph object

### PR DESCRIPTION
**Pull request by hyanwong**
To represent a immutable, valid GIG. This also includes a method to "sample resolve" a GIG, although it's not extensively tested yet.

@duncanMR - after writing the code, I think that it's easier to have child_left < child_right for inversions (rather than the same for parents), because in my `sample_resolve` code, I check the left and right in the child against the existing intervals as we go up the graph. See what you think?

There's a lot of extra functionality in this PR, sorry, but I think it's useful. I've tried to stick to the tskit style as much as possible, apart from the following:

```
gig.edges  # can iterate over, like ts.edges() but without the braces
gig.edges[0]  # same as ts.edge(0)
gig.nodes  # like ts.nodes()
gig.node[0]  # like ts.node(0)
gig.sequence_length(u)  # has to be per node: find a max sequence length for node u, using the edges
```

## Summary by CodeRabbit

- **New Features**
  - Introduced an enhanced genetic inheritance graph representation, including a new `Graph` class for efficient data manipulation.
  - Added functionality to construct genetic inheritance graphs from tree sequences.

- **Improvements**
  - Improved validation logic within the genetic inheritance graph framework for better data integrity.
  - Enhanced graph manipulation methods for more robust and versatile usage.

- **Documentation**
  - Updated test documentation to better reflect the purpose and usage of each test fixture.

- **Refactor**
  - Renamed module import alias for clarity and consistency across test files.
  - Streamlined node and edge table creation with updated references.

- **Bug Fixes**
  - Fixed issues in the genetic inheritance graph data classes to ensure immutability and correct data representation.

### CODE REVIEW

**Comment by hyanwong** [comment]
I'm please that the sample_resolve algorithm is so simple (this is the core of the simplify method), although it needs a lot more testing. I did have to use an interval library though. I does correctly sample resolve the sv gig (https://github.com/hyanwong/GeneticInheritanceGraph/issues/2#issuecomment-1858383674)

File: GeneticInheritanceGraph/graph.py [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+class Graph:
+    def __init__(self, tables):
+        """
+        Create a gig.Graph from a set of Tables. This is for internal use only:
+        the canonical way to do this is to use tables.Graph()
+        """
+        if not isinstance(tables, Tables):
+            raise ValueError("tables must be a GeneticInhertanceGraph.Tables object")
+        self.tables = tables
+        self._validate()
```
[the remark this answers is not in the document]
Comment (hyanwong) [reply GeneticInheritanceGraph/graph.py]: Done, thanks

File: GeneticInheritanceGraph/graph.py [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+                    f"Iedge {np.argmax(nodes)} contains {nm} ID >= num nodes"
+                )
+
+        # check all parents are strictly older than their children
+        node_times = self.tables.nodes.time
+        for i, ie in enumerate(self.iedges):
+            if node_times[ie.parent] <= node_times[ie.child]:
+                raise ValueError(
+                    f"Edge {i}: child node {ie.child} older than parent {ie.parent}"
+                )
```
[the remark this answers is not in the document]
Comment (hyanwong) [reply GeneticInheritanceGraph/graph.py]: done

**Review by duncanMR — APPROVED** [review]

**Comment by duncanMR** [comment]
LGTM! I haven't analysed `sample_resolve` in-depth but I'll get to that in future

File: GeneticInheritanceGraph/graph.py [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+
+        # Check that all iedges have same absolute parent span as child span
+        parent_spans = self.tables.iedges.parent_right - self.tables.iedges.parent_left
+        child_spans = self.tables.iedges.child_right - self.tables.iedges.child_left
+        span_equal = np.abs(parent_spans) == np.abs(child_spans)
+        if not np.all(span_equal):
+            bad = np.where(~span_equal)[0]
+            raise ValueError(f"iedges {bad} have different parent and child spans")
+
+        # Check that parent left is always < parent right
```
[the remark this answers is not in the document]
Comment (hyanwong) [reply GeneticInheritanceGraph/graph.py:73]: Good point, thanks

File: GeneticInheritanceGraph/graph.py [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+            raise ValueError("tables must be a GeneticInheritanceGraph.Tables object")
+        self.tables = tables
+        self._validate()
+
+    def _validate(self):
+        assert hasattr(self, "tables")
+        # Extra validation for a GIG here, e.g. edge transformations are
+        # valid, etc.
+
+        # Cached variables
```
[the remark this answers is not in the document]
Comment (hyanwong) [reply GeneticInheritanceGraph/graph.py]: Ah, I realise why I put them there: we need them defined for a tree sequence with no edges, and if there are no edges, we return from this routine just below. I have added a comment, and a test in 331bff1120dfa4818ae7c64b41c40a1e889b974c

**Comment by hyanwong** [comment]
> LGTM! I haven't analysed `sample_resolve` in-depth but I'll get to that in future

Thanks. That routine almost certainly needs more testing, but once you've had a look over it, I guess I should merge, and we can add tests later?

**Comment by hyanwong** [comment]
Note that the sample_resolve function (including the comment) is largely from https://github.com/tskit-dev/tskit/discussions/2869#discussioncomment-7553044

**Comment by duncanMR** [comment]
I cooked up a quick example with a deletion and a subtree without samples, and it behaved exactly as expected:

image

```python
import GeneticInheritanceGraph as gigl
import tests.gigutil as gigutil

    # p | c | c_left | c_right | p_left | p_right
interval_data = [
    (10, 8, 0, 3, 0, 3),
    (10, 9, 0, 5, 5, 0),
    (8, 4, 0, 3, 0, 3),
    (8, 4, 3, 6, 0, 3),
    (9, 6, 0, 5, 0, 5),
    (9, 7, 0, 5, 0, 5),
    (4, 0, 0, 6, 0, 6),
    (4, 1, 0, 6, 0, 6),
    (5, 2, 0, 3, 0, 3),
    (5, 3, 0, 3, 0, 3)
]
# time | flags
node_data = [
    (0, 1),
    (0, 1),
    (0, 1),
    (0, 1),
    (1, 0),
    (1, 0),
    (1, 0),
    (1, 0),
    (2, 0),
    (2, 0),
    (4, 0)
]

tables = gigl.Tables()
tables.intervals = gigutil.make_iedges_table(interval_data, tables)
tables.nodes = gigutil.make_nodes_table(node_data, tables)
tables.sort()
gig = gigl.Graph(tables)
resolved_gig = gig.sample_resolve()
print(gig.tables)
print(resolved_gig.tables)
```

I agree that we should merge it. It seems to work fine for simple examples; once we have a forward simulator, we'll be able to write more rigorous tests.

**Comment by hyanwong** [comment]
Great, thanks a lot for testing. I'll merge this now.

**Pull request merged**